src/python/py/models/builder.py

```diff
@@ -1018,7 +1018,104 @@ def make_rotary_embedding_caches(self, rotemb, **kwargs):
 
         return cos_cache_name, sin_cache_name
 
+
+    def make_rotary_embedding_partial(self, rotemb, name, root_input, **kwargs):
+        cos_cache_name, sin_cache_name = self.make_rotary_embedding_caches(rotemb)
```

Comment:
Partial rotary embeddings are already supported in the `RotaryEmbedding` op by using the `num_heads` and `rotary_embedding_dim` [attributes](https://github.com/microsoft/onnxruntime/blob/main/docs/ContribOperators.md#com.microsoft.RotaryEmbedding).

From reading the Hugging Face implementation of [Nemotron's rotary embeddings](https://github.com/huggingface/transformers/blob/70b07d97cf2c5f61fff55700b65528a1b6845cd2/src/transformers/models/nemotron/modeling_nemotron.py#L180-L187), the logic looks to be the same as Hugging Face's implementation of [Phi-2's rotary embeddings](https://github.com/huggingface/transformers/blob/70b07d97cf2c5f61fff55700b65528a1b6845cd2/src/transformers/models/phi/modeling_phi.py#L324-L338).

Similar to how the Phi-2 model class is created, can we enable these two attributes in the Nemotron model class to use the partial rotary embeddings instead of creating `make_rotary_embedding_partial`?

https://github.com/microsoft/onnxruntime-genai/blob/0f59a900568cf65d74554cc34027648989f4ab1c/src/python/py/models/builder.py#L2327-L2333

Reply:
I see the attributes (num_heads and rotary_embedding_dim) of RotaryEmbedding op for partial rotary factor seems to be set correctly but the output of the **phi2** and **Nemotron-Mini** both are garbage without the workaround in this change.

Since, GenAI generated RotaryEmbedding op or the implementation (DML EP), doesn't seem to be handling the partial rotary factor correctly, can we submit this as a workaround to unblock for now? And this part can be removed once a proper fix is available.

Reply:
In latest change, I have moved rotary partial implementation workaround to member function make_rotary_embedding() in Nemotron class, to minimize changes/impact. So, phi2 model behavior will not change (will be in broken state until fix is available).

Reply:
Patrice has raised PR that fixes RotaryEmbedding for partial factor in ORT, so the workaround is not needed anymore. I have now removed make_rotary_embedding override for partial factor.

https://github.com/microsoft/onnxruntime/pull/22417